coderd/prebuilds/state.go

```diff
@@ -0,0 +1,167 @@
+	"math"
+	"slices"
+	"time"
+
+	"github.com/coder/quartz"
+
+	"github.com/coder/coder/v2/coderd/database"
+)
+
+func (p PresetState) CalculateActions(clock quartz.Clock, backoffInterval time.Duration) (*ReconciliationActions, error) {
+	// TODO: align workspace states with how we represent them on the FE and the CLI
+	//	     right now there's some slight differences which can lead to additional prebuilds being created
+
+	// TODO: add mechanism to prevent prebuilds being reconciled from being claimable by users; i.e. if a prebuild is
+	// 		 about to be deleted, it should not be deleted if it has been claimed - beware of TOCTOU races!
+
+	var (
+		actual                       int32 // Running prebuilds for active version.
+		desired                      int32 // Active template version's desired instances as defined in preset.
+		eligible                     int32 // Prebuilds which can be claimed.
+		outdated                     int32 // Prebuilds which no longer match the active template version.
+		extraneous                   int32 // Extra running prebuilds for active version (somehow).
+		starting, stopping, deleting int32 // Prebuilds currently being provisioned up or down.
+	)
+
+	if p.Preset.UsingActiveVersion {
+		actual = int32(len(p.Running))
+		desired = p.Preset.DesiredInstances
+	}
+
+	for _, prebuild := range p.Running {
+		if p.Preset.UsingActiveVersion {
+			if prebuild.Ready {
+				eligible++
+			}
+
+			extraneous = int32(math.Max(float64(actual-p.Preset.DesiredInstances), 0))
+		}
+
+		if prebuild.TemplateVersionID == p.Preset.TemplateVersionID && !p.Preset.UsingActiveVersion {
+			outdated++
+		}
+	}
+
+	// In-progress builds are common across all presets belonging to a given template.
+	// In other words: these values will be identical across all presets belonging to this template.
+	for _, progress := range p.InProgress {
+		num := progress.Count
+		switch progress.Transition {
+		case database.WorkspaceTransitionStart:
+			starting += num
+		case database.WorkspaceTransitionStop:
+			stopping += num
+		case database.WorkspaceTransitionDelete:
+			deleting += num
+		}
+	}
+
+	var (
+		toCreate = int(math.Max(0, float64(
+			desired-(actual+starting)), // The number of prebuilds currently being stopped (should be 0)
+		))
+		toDelete = int(math.Max(0, float64(
+			outdated- // The number of prebuilds running above the desired count for active version
+				deleting), // The number of prebuilds currently being deleted
+		))
+
+		actions = &ReconciliationActions{
+			Actual:     actual,
+			Desired:    desired,
+			Eligible:   eligible,
+			Outdated:   outdated,
+			Extraneous: extraneous,
+			Starting:   starting,
+			Stopping:   stopping,
+			Deleting:   deleting,
+		}
+	)
+
+	// If the template has become deleted or deprecated since the last reconciliation, we need to ensure we
+	// scale those prebuilds down to zero.
+	if p.Preset.Deleted || p.Preset.Deprecated {
+		toCreate = 0
+		toDelete = int(actual + outdated)
```

Comment:
This implicitly means that we should delete all `p.Running` instances? If so maybe we can explicitly say that `toDelete=len(p.Running)`?